pkg/resource/deploy/providers/reference.go

```diff
@@ -96,6 +96,27 @@ func (r Reference) String() string {
 	return string(r.urn) + resource.URNNameDelimiter + string(r.id)
 }
 
+const denyDefaultProviderID resource.ID = "denydefaultprovider"
+
+// DenyDefaultProvider represent a default provider that cannot be created.
+func NewDenyDefaultProvider(name tokens.QName) Reference {
+	return mustNewReference(
+		resource.NewURN("denied", "denied", "denied", "pulumi:providers:denied", name),
+		denyDefaultProviderID)
+}
+
+// Retrieves the name of the denied provider.
+//
+// Panics if called on a provider that is not a DenyDefaultProvider.
+func DeniedDefaultProvider(ref Reference) string {
```

Comment:
```suggestion
func GetDeniedDefaultProviderName(ref Reference) string {
```
I wonder if this should just be `GetProviderName` and allow any provider ref? Or maybe just use ref.URN().Name() inline?

Reply:
It really should be (and now is) called `GetDeniedDefaultProviderPkg`. I'm just storing the desired property in the `ref.URN().Name()` field. I clarified the intended behavior in comments.